k8sattributes processor documentation PR

Yadumathur opened the pull request:
I have opened this PR with respect to the following issue which I have raised 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7912

Below are the addition in the current documentation 
Daemon set :
1. Adding the env variable for POD IP in downward api for Daemon set ( config example ) 
2. OTEL_RESOURCE_ATTRIBUTES config to update the pod ip 
Config Map 
1. Adding resourcedetectionprocessor to the config map config example 
2. Adding the k8sattributes config to extract k8s meta data

**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

Yadumathur committed:
k8sattributes processor documentation PR

I have opened this PR with respect to the following issue which I have raised 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7912

Below are the addition in the current documentation 
Daemon set :
1. Adding the env variable for POD IP in downward api for Daemon set ( config example ) 
2. OTEL_RESOURCE_ATTRIBUTES config to update the pod ip 
Config Map 
1. Adding resourcedetectionprocessor to the config map config example 
2. Adding the k8sattributes config to extract k8s meta data

Yadumathur committed:
Update doc.go

dmitryax commented on processor/k8sattributesprocessor/doc.go:
```diff
@@ -117,12 +141,22 @@
+//            fieldRef:
+//             apiVersion: v1
+//              fieldPath: spec.nodeName
+//        - name: POD_IP
+//          valueFrom:
+//            fieldRef:
+//              apiVersion: v1
+//              fieldPath: status.podIP
+//        - name: OTEL_RESOURCE_ATTRIBUTES
+//          value: k8s.pod.ip=$(POD_IP)
```
`k8s.pod.ip` is supposed to be IP of a pod emitting data, not the collector itself. It's usually taken from the context of tracing data.

So this is correct configuration part that can be set on **instrumented pod**, not on the cllector

Yadumathur committed:
Updated a new section with "Optional" config

Yadumathur committed:
Update doc.go

pmm-sumo commented on processor/k8sattributesprocessor/doc.go:
```diff
@@ -94,7 +94,31 @@
 //
-// TODO: example config.
+//   processors:
+//      resourcedetection:
```
Why is this processor included here?

pmm-sumo reviewed (changes requested):
I must be missing something but I don't know how this would work. 

In general, it's a viable solution when instrumentation libraries are used (so the local environment variable is pulled and used in the exported data). 

However, with filelog receiver scenario it's much different. It typically [mounts the varlog](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/examples/kubernetes/otel-collector.yaml#L113-L115) and tails the content locally. Which means it does not have access to env variables of the respective pods

Yadumathur committed:
removed resourcedetection stanza

Yadumathur committed:
removed pod level env

Yadumathur committed:
removed pod level env

Yadumathur committed:
Update processor/k8sattributesprocessor/doc.go

Co-authored-by: Przemek Maciolek <58699843+pmm-sumo@users.noreply.github.com>

pmm-sumo commented on processor/k8sattributesprocessor/doc.go:
```diff
@@ -94,7 +94,28 @@
+//         - from: resource_attribute
+//           name: k8s.pod.ip
+//         - from: resource_attribute
+//           name: k8s.pod.uid
```
I think it would be good to include connection association in the example, since it's fairly commonly used for tracing
```suggestion
//           name: k8s.pod.uid
//         - from: connection
//           name: ip
```

pmm-sumo reviewed (approved):
I think it looks OK now (though does not resolve the original problem with logs). What do you think @dmitryax ?

Yadumathur committed:
Update processor/k8sattributesprocessor/doc.go

Agreed , it should be part of the doc

Co-authored-by: Przemek Maciolek <58699843+pmm-sumo@users.noreply.github.com>

Yadumathur reviewed (commented):
I have made changes which were asked for

Yadumathur commented:
> I think it looks OK now (though does not resolve the original problem with logs). What do you think @dmitryax ?

Yadumathur commented:
I was not able to add "Skip Changelog" label, looks like closed it 
@pmm-sumo  can you help me in reopening this PR and add skip changelog label ?

dmitryax commented on processor/k8sattributesprocessor/doc.go:
```diff
@@ -117,15 +140,19 @@
 //
 // This will inject a new environment variable to the OpenTelemetry container with the value as the
 // name of the node the pod was scheduled to run on.
+
```
Is this line added accidentally?

Yadumathur replied:
No It was already there, This is a part of Otel C container spec.nodeName which we are setting up as KUBE_NODE_NAME and using it in the k8sattributes filter to make sure we use fewer resources.
It was there in the previous version of the document as well.

dmitryax replied:
I mean the empty line added with this PR. Looks unrelated

Yadumathur replied:
seems that was added accidentally , I have removed it

dmitryax reviewed (commented):
LGTM. just one nit

Yadumathur committed:
removed empty line

dmitryax committed:
Update doc.go

Yadumathur committed:
Updating the K8s otel-collector.yaml file  

K8sprocessor uses resource level attribute to fetch k8s metadata using k8sattributeprocessor.
Hence adding the code part in the example config for users to look at, also refers to the issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7912

pmm-sumo commented on examples/kubernetes/otel-collector.yaml:
```diff
@@ -64,20 +64,58 @@ data:
+            - tag_name: k8s.label.component
+              key: component
+              from: pod
+#Pod association using resource attribute 
```
```suggestion
      # pod association using resource attribute 
```

Yadumathur commented:
@pmm-sumo and @dmitryax Can you please review one more file which I have updated? I know you guys already approved previous changes. -- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/examples/kubernetes/otel-collector.yaml

The main important part of pod association is the resource level attribute and hence when you  move out attributes to Attributes , your metadata should also tag resource which we need for pod association in config as below
            resource:
              k8s.pod.uid: 'EXPR($.uid)'

Hence , I have updated the example kubernetes file which most users will fall back on for better understanding of whole set up 
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/examples/kubernetes/otel-collector.yaml
Let me know this works or not ?

Yadumathur committed:
Update examples/kubernetes/otel-collector.yaml

Co-authored-by: Przemek Maciolek <58699843+pmm-sumo@users.noreply.github.com>

Yadumathur committed:
Update examples/kubernetes/otel-collector.yaml

Co-authored-by: Przemek Maciolek <58699843+pmm-sumo@users.noreply.github.com>

sumo-drosiek commented on examples/kubernetes/otel-collector.yaml:
```diff
@@ -64,20 +64,58 @@ data:
               k8s.pod.uid: 'EXPR($.uid)'
+          # tag the resource level attribute in metadata
+            resource:
+              k8s.pod.uid: 'EXPR($.uid)'
```
Maybe something about grouping by the resource attributes?
I don't remember :shrug:

dmitryax replied:
I agree that `k8s.*` attributes should go to resource

Yadumathur replied:
agree ,

pmm-sumo replied:
I tested it and it seems moving all attributes to the resource worked as anticipated (on collector version v0.45.0):

```
Resource SchemaURL:
Resource labels:
     -> k8s.container.restart_count: STRING(0)
     -> k8s.namespace.name: STRING(kube-system)
     -> k8s.pod.name: STRING(kube-controller-manager-docker-desktop)
     -> k8s.pod.uid: STRING(e569b29c778f62de29f383a9cd6709a9)
     -> k8s.container.name: STRING(kube-controller-manager)
InstrumentationLibraryLogs #0
InstrumentationLibraryLogs SchemaURL:
InstrumentationLibrary
LogRecord #0
Timestamp: 2022-02-22 16:45:56.165959811 +0000 UTC
Severity:
ShortName:
Body: W0222 16:45:56.165662       1 somefile.go:306] Error ....

Attributes:
     -> stream: STRING(stderr)
   ```

Yadumathur replied:
Yeah. it works, 
tested on my end as well

dmitryax committed:
Update otel-collector.yaml

dmitryax committed:
Update otel-collector.yaml

dmitryax commented on examples/kubernetes/otel-collector.yaml:
```diff
@@ -64,20 +64,57 @@ data:
+          - from: resource_attribute
+            name: k8s.pod.ip
+          - from: connection
+            name: ip
```
I don't think `connection` uses `name` field

pmm-sumo replied:
BTW, we might want to improve the processor docs. I thought that keeping "name" there was used as a convention :)

Yadumathur replied:
yeah, I missed looking at the code of this pod association. I was not using it as is just kept it consistent. 
Lets update this one as well correctly

pmm-sumo commented:
I think it would be worthwhile to update the image version in otel-collector.yaml (currently it's `image: otel/opentelemetry-collector-contrib:0.37.1`)

dmitryax commented on examples/kubernetes/otel-collector.yaml:
```diff
@@ -64,20 +64,57 @@ data:
+            - k8s.pod.start_time
+          # Pod labels which can be fetched via K8sattributeprocessor
+          labels:
+            - tag_name: k8s.label.app
```
`k8s.labels.app` is default I believe. Let's update it to default or remove the field.

Yadumathur replied:
Just made a change to make it generic

dmitryax commented on examples/kubernetes/otel-collector.yaml:
```diff
@@ -64,20 +64,57 @@ data:
+            - tag_name: k8s.label.app
+              key: app
+              from: pod
+            - tag_name: k8s.label.component
```
Same here

dmitryax commented:
I pushes few formatting fixes to avoid review back-and-forths

Yadumathur committed:
fixed the changes to be more generic

pmm-sumo commented:
BTW, I played with the manifest file and the current regex did not work:

```
2022-02-22T16:49:34.493Z	error	Failed to process entry	{"kind": "receiver", "name": "filelog", "operator_id": "$.extract_metadata_from_filepath", "operator_type": "regex_parser", "error": "regex pattern does not match", "action": "send", "entry": {"timestamp":"2022-02-22T16:49:34.298896618Z","body":{"log":"2022-02-22 16:49:34.298613 I | etcdserver/api/etcdhttp: /health OK (status code 200)\n","stream":"stderr"},"attributes":{"file.path":"/var/log/pods/kube-system_etcd-docker-desktop_5d9d97b8d8daed31d6fd5c6d386c29c5/etcd/0.log"},"severity":0}}
```

Yadumathur commented:
> BTW, I played with the manifest file and the current regex did not work (K8s 1.22.5).
> 
> ```
> 2022-02-22T16:49:34.493Z	error	Failed to process entry	{"kind": "receiver", "name": "filelog", "operator_id": "$.extract_metadata_from_filepath", "operator_type": "regex_parser", "error": "regex pattern does not match", "action": "send", "entry": {"timestamp":"2022-02-22T16:49:34.298896618Z","body":{"log":"2022-02-22 16:49:34.298613 I | etcdserver/api/etcdhttp: /health OK (status code 200)\n","stream":"stderr"},"attributes":{"file.path":"/var/log/pods/kube-system_etcd-docker-desktop_5d9d97b8d8daed31d6fd5c6d386c29c5/etcd/0.log"},"severity":0}}
> ```
> 
> And the reason is that uid should be 32 characters rather [than 36](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7913/files#diff-47e8be1302ed7f3f501f789875147072895ef72f4ed7cc615e572b7459021a69L56): `^.*\/(?P<namespace>[^_]+)_(?P<pod_name>[^_]+)_(?P<uid>[a-f0-9\-]{32})\/(?P<container_name>[^\._]+)\/(?P<restart_count>\d+)\.log$`
> 
> Or even better:
> 
> ```
> `^.*\/(?P<namespace>[^_]+)_(?P<pod_name>[^_]+)_(?P<uid>[a-f0-9\-]{32,36})\/(?P<container_name>[^\._]+)\/(?P<restart_count>\d+)\.log$`
> ```

I was already using ^.*\/(?P<namespace>[^_]+)_(?P<pod_name>[^_]+)_(?P<uid>[a-f0-9\-]{36})\/(?P<container_name>[^\._]+)\/(?P<restart_count>\d+)\.log$ hence never faced issue

Yadumathur committed:
connection does not uses asso.Name

dmitryax committed:
Update doc.go

dmitryax committed:
Update otel-collector.yaml

pmm-sumo commented:
> I was already using ^.*/(?P[^_]+)_(?P<pod_name>[^_]+)_(?P[a-f0-9-]{36})/(?P<container_name>[^\._]+)/(?P<restart_count>\d+).log$ hence never faced issue

UUID's can be expressed with or without dashes which makes both 32 and 36 characters valid. I think the example should be robust enough to cover both so I suggest to use {32,36} here

_(Apologies, for whatever reason I clicked the wrong button and thought I'm replying while I edited the original message, so let me rewrite my message :) )_
